src/platform/cmx/chal.c

```diff
@@ -0,0 +1,205 @@
+				RASR|=(((u32_t)0x0F)<<(count*4+8));
+		}
+	}
+	else if(num_order==2)
+	{
+		for(count=0;count<2;count++)
+		{
+			if(((table->data[count]&PGTBL_PRESENT)!=0)&&((table->data[count]&PGTBL_LEAF)!=0))
+				RASR|=(((u32_t)0x03)<<(count*2+8));
+		}
+	}
+	else
+	{
+		for(count=0;count<8;count++)
+		{
+			if(((table->data[count]&PGTBL_PRESENT)!=0)&&((table->data[count]&PGTBL_LEAF)!=0))
+				RASR|=(((u32_t)1)<<(count+8));
+		}
+	}
+	if(RASR==0)
+		return 0;
+
+	RASR=CMX_MPU_SRDCLR&(~RASR);
+	RASR|=CMX_MPU_SZENABLE;
+	/* Because what we type is always read-write, thus the RASR is always RW */
+	RASR|=CMX_MPU_RW;
+	/* Can we fetch instructions from there? - always cannot, we do not allow data execution */
+	RASR|=CMX_MPU_XN;
+	/* Is the area cacheable? - because what we type is SRAM, so it is always accessible */
+	RASR|=CMX_MPU_CACHEABLE;
+	/* Is the area bufferable? - always bufferable */
+	RASR|=CMX_MPU_BUFFERABLE;
+	/* What is the region size? */
+	RASR|=CMX_MPU_REGIONSIZE(size_order, num_order);
+        /* 1303fc1b */
+	return RASR;
+}
+
+u32_t
+chal_MPU_clear(struct pgtbl* table, u32_t start_addr, u32_t size_order)
+{
+	u32_t count;
+
+	/* There are only 4 regions at the toplevel which is configurable */
+	for(count=0;count<4;count++)
+	{
+		if((table->data[9]&(((u32_t)1)<<count))!=0)
+		{
+			/* We got one MPU region valid here */
+			if((CMX_MPU_ADDR(table->data[count*2])==start_addr)&&
+			   (CMX_MPU_SZORD(table->data[count*2+1])==size_order))
+			{
+				/* Clean it up and return */
+				table->data[count*2]=CMX_MPU_VALID|count;
+				table->data[count*2+1]=0;
+				table->data[9]&=~(((u32_t)1)<<count);
+				return 0;
+			}
+		}
+	}
+
+	return 0;
+}
+
+u32_t
+chal_MPU_add(struct pgtbl* table, u32_t start_addr, u32_t size_order, u32_t MPU_RASR)
+{
+	u32_t count;
+	/* The last empty slot */
+	u32_t last_empty;
+
+	/* Set these values to some overrange value */
+	last_empty=16;
+
+	/* We only set 4 regions */
+	for(count=0;count<4;count++)
+	{
+		if((table->data[9]&(((u32_t)1)<<count))!=0)
+		{
+			/* We got one MPU region valid here */
+			if((CMX_MPU_ADDR(table->data[count*2])==start_addr)&&   /* RBAR */
+			   (CMX_MPU_SZORD(table->data[count*2+1])==size_order)) /* RASR */
+			{
+				/* Update the RASR - all flag changes except static are reflected here */
+				table->data[count*2+1]=MPU_RASR;
+				return 0;
+			}
+		}
+		else
+			last_empty=count;
+	}
+
+	/* Update unsuccessful, we didn't find any match. We will need a new slot.
+	 * See if there are any new empty slots that we can use.
+	 * See if the last empty slot is 0. If yes, we can only map dynamic pages */
+	if(last_empty!=16)
+	{
+		/* Put the data to this slot */
+		table->data[9]|=((u32_t)1)<<(last_empty);
+		table->data[last_empty*2]=CMX_MPU_ADDR(start_addr)|CMX_MPU_VALID|last_empty;
+		table->data[last_empty*2+1]=MPU_RASR;
+
+		return 0;
+	}
+
+	/* All effort is futile. we report failure */
+	return -1;
+}
+
+u32_t chal_MPU_update(struct pgtbl* table, u32_t Op_Flag)
```

Comment:
This function will update the current MPU settings with the page table contents.